### PR TITLE
Queues Laura and Natalia Scrabble Game

### PR DESCRIPTION
**Pull request by natalia-ku**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	|   We worked excellent together thru the project. We switched between driver and navigator very often, and helped each other. 	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | We don't need an instance of Scoring class , we can call score method without creating Scoring class instance    |
|   Describe an example of a test you wrote for a _nominal case_. |  test score(word) method for words that have from 2 to 6 letters:  Scrabble::Scoring.score("com").must_equal 7	|
|   Describe an example of a test you wrote for an _edge case_. |  test score(word)  method with word that has 1 letter, or with word that has 7 letters : Scrabble::Scoring.score("H").must_equal 4 Scrabble::Scoring.score("jacuzzi").must_equal 84	|
|   What was your final test coverage percentage?	|   100%	|
|   Is there a specific piece of code you'd like feedback on?	|  1. Why simplecov behave so weirdly? 	2. Whatever you think we could make cleaner in our code?|

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
+module Scrabble
+  class Scoring
+    LETTERS_VALUES = {
```
Comment (droberts-sea) [inline lib/scoring.rb:3]: This hash of arrays definitely works, but looking up a letter's score ends up being complicated, involving an `each` loop.

Instead, what if you switched it around and had the keys in this array be letters, and the values be the corresponding score. So it would look like

```ruby
LETTERS_VALUES = {
  'a': 1,
  'b': 3,
  # ...
}
```

Then looking up a letter would be as simple as `LETTERS_VALUES[letter]`.

File: specs/scoring_spec.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require 'minitest/skip_dsl'
+require_relative '../lib/scoring'
+Minitest::Reporters.use!
+
+describe "Scoring class" do
+  it "Creates new instance of Scoring class" do
+    Scrabble::Scoring.new.class.must_equal Scrabble::Scoring
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:12]: I'm not sure if this test is necessary - since you're not supposed to create an instance of `Scrabble::Scoring`, it doesn't make sense to test that behavior.

File: lib/scoring.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      1 => %w(A E I O U L N R S T),
+      2 => %w(D G),
+      3 => %w(B C M P),
+      4 => %w(F H V W Y),
+      5 => %w(K),
+      8 => %w(J X),
+      10 => %w(Q Z)
+    }
+
+    def self.input_correct?(word)
```
Comment (droberts-sea) [inline lib/scoring.rb:13]: Good work abstracting this behavior out to its own method.

File: specs/player_spec.rb [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+      player.play("hello")
+      player.won?.must_equal true
+    end
+
+    it "Returns false if total score < 100" do
+      player = Scrabble::Player.new("Natalia")
+      player.play("flower")
+      player.play("hello")
+      player.won?.must_equal false
+    end
```
Comment (droberts-sea) [inline specs/player_spec.rb:90]: I would like to see some testing around the boundary here - check it when the score is 99 and 100, and make sure it behaves as expected.

File: lib/tile_bag.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  class TileBag
+    attr_reader :tile_bag
+
+    def initialize
+      @tile_bag = Scrabble::TileBag.create_tile_bag
+    end
+
+    def self.create_tile_bag
+      letters = ("a".."z").to_a
+      quantity= [9,2,2,4,12,2,3,2,9,1,1,4,2,6,8,2,1,6,4,6,4,2,2,1,2,1]
```
Comment (droberts-sea) [inline lib/tile_bag.rb:11]: I like the idea here, but this representation is hard to read / modify. What if you needed to change the number of "N"s - you'd have to count to figure out which one it was.

In this case, it might be better to be more verbose, something like

```ruby
tile_bag = []
tile_bag += ['A'] * 9
tile_bag += ['B'] * 2
# ... etc ...
```

File: lib/tile_bag.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      tile_bag = []
+
+      26.times do |i|
```
Comment (droberts-sea) [inline lib/tile_bag.rb:14]: Having that `26` sitting around makes me nervous - why not use `letters.length.times` instead?

File: lib/tile_bag.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    def draw_tiles(num)
+      if num.class != Integer
+        raise ArgumentError.new("Parameter must be integer")
+      end
+
+      drawn_tiles = []
+      num.times do
+        rand_num = rand(0...@tile_bag.length)
+        drawn_tiles << @tile_bag[rand_num]
+        @tile_bag.delete_at(rand_num)
```
Comment (droberts-sea) [inline lib/tile_bag.rb:31]: Good solution to this problem. A slightly cleaner way might be to use `Array#shuffle` and `Array#pop` - go look them up!

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|   yes	|
|   Both Teammates contributed to the codebase	|   yes	|
|   Regular Commits with meaningful commit messages 	|   yes	|
|  **Scoring class**	|   	|
|  `score` method	|   yes	|
|   Uses appropriate data structure to store the letter score	|   see inline comment	|
|   Appropriately handles edge cases	|   yes	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|   yes	|
|  `highest_scoring_word` method	|   yes	|
|   Appropriately handles edge cases	|   yes	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|   yes (via tie and find_max)	|
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|   yes	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   yes	|
|  Returns `true` or `false` from `won?` method |  yes |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |    yes |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|   see inline comment	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |   yes	|
| `tiles_remaining` method returns a count	|   yes	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   yes	|

Great work overall. Code is for the most part clean and readable, and git hygiene looks good. Keep up the good work.

The one thing I'd like to see is more deliberate choice of data structures. In both `Score` and `TileBag`, you used a representation that ended up being hard to work with - make sure you do your research and really think your choices through.
